File: backend-csharp/Services/LoginHistoryService.cs
using Microsoft.EntityFrameworkCore;
using Picture2Text.Api.Data;
using Picture2Text.Api.Models;

namespace Picture2Text.Api.Services;

/// <summary>
/// 登入歷史記錄服務
/// </summary>
public class LoginHistoryService
{
    private readonly ApplicationDbContext _context;
    private readonly ILogger<LoginHistoryService> _logger;

    public LoginHistoryService(
        ApplicationDbContext context,
        ILogger<LoginHistoryService> logger)
    {
        _context = context;
        _logger = logger;
    }

    /// <summary>
    /// 記錄登入嘗試
    /// </summary>
    public async Task RecordLoginAttemptAsync(
        string attemptedUserId,
        int? userId,
        bool isSuccess,
        string? failureReason,
        string? ipAddress,
        string? userAgent,
        string? deviceInfo = null)
    {
        // 記錄結構化日誌
        if (isSuccess)
        {
            _logger.LogInformation(
                "用戶登入成功 - UserId: {UserId}, AttemptedUserId: {AttemptedUserId}, IP: {IpAddress}",
                userId, attemptedUserId, ipAddress);
        }
        else
        {
            _logger.LogWarning(
                "用戶登入失敗 - AttemptedUserId: {AttemptedUserId}, IP: {IpAddress}, Reason: {FailureReason}",
                attemptedUserId, ipAddress, failureReason);
        }

        var history = new LoginHistory
        {
            UserId = userId,
            AttemptedUserId = attemptedUserId,
            IsSuccess = isSuccess,
            FailureReason = failureReason,
            IpAddress = ipAddress,
            UserAgent = userAgent,
            DeviceInfo = deviceInfo,
            AttemptedAt = DateTime.UtcNow
        };

        _context.LoginHistories.Add(history);
        await _context.SaveChangesAsync();
        
        _logger.LogDebug(
            "登入歷史已儲存 - LoginHistoryId: {LoginHistoryId}, Success: {IsSuccess}",
            history.Id, isSuccess);
    }

    /// <summary>
    /// 取得使用者的登入歷史
    /// </summary>
    public async Task<List<LoginHistory>> GetUserLoginHistoryAsync(int userId, int limit = 50)
    {
        _logger.LogDebug("查詢用戶登入歷史 - UserId: {UserId}, Limit: {Limit}", userId, limit);
        
        var histories = await _context.LoginHistories
            .Where(h => h.UserId == userId)
            .OrderByDescending(h => h.AttemptedAt)
            .Take(limit)
            .ToListAsync();
            
        _logger.LogInformation(
            "查詢完成 - UserId: {UserId}, 找到 {Count} 筆記錄", 
            userId, histories.Count);
            
        return histories;
    }

    /// <summary>
    /// 取得最近的失敗登入嘗試次數（用於防暴力破解）
    /// </summary>
    public async Task<int> GetRecentFailedAttemptsAsync(string attemptedUserId, int minutesWindow = 15)
    {
        var windowStart = DateTime.UtcNow.AddMinutes(-minutesWindow);
        
        var failedAttempts = await _context.LoginHistories
            .Where(h => h.AttemptedUserId == attemptedUserId 
                     && !h.IsSuccess 
                     && h.AttemptedAt >= windowStart)
            .CountAsync();
            
        if (failedAttempts > 5)
        {
            _logger.LogError(
                "⚠️ 偵測到異常登入行為！AttemptedUserId: {AttemptedUserId} 在 {MinutesWindow} 分鐘內失敗 {FailureCount} 次",
                attemptedUserId, minutesWindow, failedAttempts);
        }
        else if (failedAttempts > 3)
        {
            _logger.LogWarning(
                "注意：AttemptedUserId: {AttemptedUserId} 在 {MinutesWindow} 分鐘內失敗 {FailureCount} 次",
                attemptedUserId, minutesWindow, failedAttempts);
        }
        
        return failedAttempts;
    }

    /// <summary>
    /// 清理舊的登入歷史記錄
    /// </summary>
    public async Task CleanupOldHistoryAsync(int daysToKeep = 90)
    {
        _logger.LogInformation("開始清理舊的登入歷史記錄 - 保留天數: {DaysToKeep}", daysToKeep);
        
        var cutoffDate = DateTime.UtcNow.AddDays(-daysToKeep);
        
        var oldRecords = await _context.LoginHistories
            .Where(h => h.AttemptedAt < cutoffDate)
            .ToListAsync();

        if (oldRecords.Any())
        {
            _logger.LogInformation(
                "正在刪除 {RecordCount} 筆舊記錄（早於 {CutoffDate:yyyy-MM-dd}）",
                oldRecords.Count, cutoffDate);
                
            _context.LoginHistories.RemoveRange(oldRecords);
            await _context.SaveChangesAsync();
            
            _logger.LogInformation("清理完成 - 已刪除 {RecordCount} 筆記錄", oldRecords.Count);
        }
        else
        {
            _logger.LogInformation("沒有需要清理的舊記錄");
        }
    }
}

File: backend-nestjs/src/services/login-history.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, LessThan, MoreThan, And } from 'typeorm';
import { LoginHistory } from '../entities';
import { LoginHistoryInfo } from '../dto/response';

@Injectable()
export class LoginHistoryService {
  private readonly logger = new Logger(LoginHistoryService.name);

  constructor(
    @InjectRepository(LoginHistory)
    private readonly loginHistoryRepository: Repository<LoginHistory>,
  ) {}

  /**
   * 記錄登入嘗試
   */
  async recordLoginAttemptAsync(
    attemptedUserId: string,
    isSuccess: boolean,
    failureReason?: string,
    userId?: number,
    ipAddress?: string,
    userAgent?: string,
    deviceInfo?: string,
    location?: string,
  ): Promise<LoginHistory> {
    const history = this.loginHistoryRepository.create({
      UserId: userId ?? null,
      AttemptedUserId: attemptedUserId,
      IsSuccess: isSuccess,
      FailureReason: failureReason ?? null,
      IpAddress: ipAddress ?? null,
      UserAgent: userAgent ?? null,
      DeviceInfo: deviceInfo ?? null,
      AttemptedAt: new Date(),
      Location: location ?? null,
    });

    await this.loginHistoryRepository.save(history);

    if (isSuccess) {
      this.logger.log(`Successful login attempt for user ${attemptedUserId} from ${ipAddress}`);
    } else {
      this.logger.warn(`Failed login attempt for user ${attemptedUserId} from ${ipAddress}: ${failureReason}`);
    }

    return history;
  }

  /**
   * 取得使用者的登入歷史記錄
   */
  async getUserLoginHistoryAsync(userId: number, limit = 20): Promise<LoginHistoryInfo[]> {
    const histories = await this.loginHistoryRepository.find({
      where: { UserId: userId },
      order: { AttemptedAt: 'DESC' },
      take: limit,
    });

    return histories.map((history) => ({
      Id: history.Id,
      IsSuccess: history.IsSuccess,
      FailureReason: history.FailureReason || undefined,
      IpAddress: history.IpAddress || undefined,
      DeviceInfo: history.DeviceInfo || undefined,
      Location: history.Location || undefined,
      AttemptedAt: history.AttemptedAt,
    }));
  }

  /**
   * 取得最近的失敗登入嘗試次數（用於暴力破解防護）
   */
  async getRecentFailedAttemptsAsync(
    attemptedUserId: string,
    windowMinutes = 15,
  ): Promise<number> {
    const windowStart = new Date();
    windowStart.setMinutes(windowStart.getMinutes() - windowMinutes);

    const count = await this.loginHistoryRepository.count({
      where: {
        AttemptedUserId: attemptedUserId,
        IsSuccess: false,
        AttemptedAt: MoreThan(windowStart),
      },
    });

    return count;
  }

  /**
   * 檢查帳戶是否被鎖定
   */
  async isAccountLockedAsync(
    attemptedUserId: string,
    maxFailedAttempts = 5,
    windowMinutes = 15,
  ): Promise<boolean> {
    const failedAttempts = await this.getRecentFailedAttemptsAsync(attemptedUserId, windowMinutes);
    return failedAttempts >= maxFailedAttempts;
  }

  /**
   * 清理舊的登入歷史記錄
   */
  async cleanupOldHistoryAsync(retentionDays = 90): Promise<number> {
    const cutoffDate = new Date();
    cutoffDate.setDate(cutoffDate.getDate() - retentionDays);

    const result = await this.loginHistoryRepository.delete({
      AttemptedAt: LessThan(cutoffDate),
    });

    const count = result.affected || 0;
    if (count > 0) {
      this.logger.log(`Cleaned up ${count} login history records older than ${retentionDays} days`);
    }

    return count;
  }
}
